action-server/covidflow/actions/form_helper.py

```diff
@@ -9,28 +9,74 @@
 logger = structlog.get_logger()
 
 
+VALIDATION_FUNCTION_PREFIX = "validate_"
+
+
 def request_next_slot(
     form: FormAction,
     dispatcher: CollectingDispatcher,
     tracker: Tracker,
     domain: Dict[Text, Any],
-    utter_ask_slot_template: Callable[[str, Tracker], Optional[str]],
+    utter_ask_slot_template: Optional[Callable[[str, Tracker], Optional[str]]] = None,
 ) -> Optional[List[EventType]]:
     """Request the next slot and utter template if needed,
-        else return None"""
+        else return None.
+        If requested doesn't change (we didn't get a valid answer), plays error message if present in domain
+        This behaviour is overridable with utter_ask_slot_template parameter
+        """
 
     for slot in form.required_slots(tracker):
         if form._should_request_slot(tracker, slot):
             logger.debug(f"Request next slot '{slot}'")
 
-            template = utter_ask_slot_template(slot, tracker)
+            template = (
+                utter_ask_slot_template(slot, tracker)
+                if utter_ask_slot_template
+                else None
+            )
 
             if template is None:
-                template = f"utter_ask_{slot}"
+                if tracker.get_slot(REQUESTED_SLOT) == slot and _has_template(
+                    domain, f"utter_ask_{slot}_error"
+                ):
+                    template = f"utter_ask_{slot}_error"
+                else:
+                    template = f"utter_ask_{slot}"
 
             dispatcher.utter_message(template=template, **tracker.slots)
 
             return [SlotSet(REQUESTED_SLOT, slot)]
 
     # no more required slots to fill
     return None
+
+
+def yes_no_nlu_mapping(form: FormAction) -> List[Dict]:
+    return [
+        form.from_intent(intent="affirm", value=True),
+        form.from_intent(intent="deny", value=False),
+        form.from_text(),
+    ]
+
+
+def validate_boolean_slot(validator):
+    def new_validator(
+        thing,
```

Comment:
I'm sure you can find something beter ;)

Reply:
XD

Reply:
oups

Reply:
`self` would work, AFAIK